Add more timer tests

mark0n opened the pull request:
The existing tests are testing many things in each test. In contrast to that the tests I'm adding here are each testing one basic property (they can also act as a sort of documentation). Each test is designed to fail for one reason only - creating a fresh timer queue for each test helps with isolation.

Note that currently these tests are leveraging C++11 features. We could either

- merge these tests into a new 7.1 branch (which will be the first EPICS version which requires a C++11 compiler)
- refactor them to make them compatible with C++98 for now (and then refactor them for better readability later when we are working on 7.1)
- put `ifdef`s around the new code to include these tests only with compilers which support C++11

Which one do you prefer?

mark0n committed:
Add more tests

anjohnson commented:
The 3.15 branch doesn't provide `epicsTime::getMonotonic()`, and is never going to require C++11, so this PR seems to be targeting the wrong branch.

I don't have a problem with adding optional tests that require specific compiler features to run. The `epicsUnitTest.h` API even provides a placeholder for them so you don't have to adjust the number of tests in the plan, that's what `testSkip()` is for – described in the comments in that header file.

mark0n commented:
I replaced `getMonotonic()` by `getCurrent()`. With this change this PR can indeed go into the 3.15 branch. The new tests will now be skipped automatically when the compiler doesn't support C++11.

The failing tests on Ubuntu are due to #106, they should pass once #107 is merged.

mark0n commented:
The TAP file shows it is skipping my new tests on MacOS. Shouldn't clang 12 on MacOS support C++11?

mark0n committed:
Allow MSVC to report a correct value for __cplusplus

Although MSVC supports most features of modern C++ for whatever
reason the __cplusplus macro kept reporting "199711L". With MSVC 19
the macro finally provides a reasonable value but only when
we pass a command-line parameter.

mark0n commented:
This is ready for another round of review. Note that the timer tests sporadically failing on MacOS are the _old_ tests. The new tests are skipped on MacOS since the `__cplusplus` macro doesn't report compatibility with C++11 (can someone with a little more Mac experience please confirm that that's correct?).

mdavidsaver commented on configure/os/CONFIG.win32-x86.win32-x86:
```diff
@@ -185,7 +185,9 @@ RES=.res
-COMPILER_CXXFLAGS = -TP
+# We want MSVC to report a reasonable C++ version in the __cplusplus macro so
+# we need to set -Zc:__cplusplus (this is at least required for MSVC 19.xx).
+COMPILER_CXXFLAGS = -TP -Zc:__cplusplus
```
I hadn't come across the [`-Zc:*` family of options](https://docs.microsoft.com/en-us/cpp/build/reference/zc-conformance?view=msvc-160) before.  I had come across [`/std=*`](https://docs.microsoft.com/en-us/cpp/build/reference/std-specify-language-standard-version?view=msvc-160), but that is >=2017.

[-Zc:__cplusplus](https://docs.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=msvc-160) is documented as >=2017 (15.7).  Since the older appveyor builds succeed without additional warnings, I guess the older compilers are silently ignoring unknown `-Zc:*` operations?

And hey, with [`/Zc:preprocessor`](https://docs.microsoft.com/en-us/cpp/build/reference/zc-preprocessor?view=msvc-160) the MS preprocessor is finally come up to [C99 conformance](https://devblogs.microsoft.com/cppblog/msvc-preprocessor-progress-towards-conformance/).

mdavidsaver commented on src/libCom/test/epicsTimerTest.cpp:
```diff
@@ -27,6 +35,150 @@
+  std::function<void()> expireFct;
+};
+
+void verifyExpirationTime(double delta_t, double tolerance) {
```
`delta_t` is slightly confusion to me.   In most c++ code I've seen, `*_t` is a type.

mark0n replied:
:+1: Fixed.

mdavidsaver commented on src/libCom/test/epicsTimerTest.cpp:
```diff
@@ -27,6 +35,150 @@
+};
+
+void verifyExpirationTime(double delta_t, double tolerance) {
+  if (!testOk(std::abs(delta_t) < tolerance, "timer expired at the expected "
```
This tight unsigned test seems problematic.  Based on discussions elsewhere, I think we are in agreement that this delay should never by negative?  At the same time, on shared CI builder services, it could easily be longer.  Would a better pass condition be: `delta >= 0.0 && (testImpreciseTiming() || delta < tolerance)`?

mdavidsaver replied:
Also, to help with troubleshooting CI build failures, tests like this one should print the values being tested.

mdavidsaver replied:
... which I just noticed is being done on the next line (which was cropped out).

anjohnson replied:
This PR is for 3.15, which doesn't have `testImpreciseTiming()`.

mdavidsaver replied:
Shouldn't this addition be going into 7.0?

mark0n replied:
Done. The 8 failing tests on Linux are due to #106 and will go away when #107 gets merged. For now I didn't include the `testImpreciseTiming()` part to maintain compatibility with 3.15. I'm happy to add that once this change has been merged up to 7.0.

mdavidsaver commented on src/libCom/test/epicsTimerTest.cpp:
```diff
@@ -27,6 +35,150 @@
+    const double arbitrary_time { 0.3 };
+    h1.start(arbitrary_time);
+
+    epicsThreadSleep(arbitrary_time + 0.1);
```
I'm not enthused by all of the sleeps, especially with hard-coded delays.  I would like to see all of these tests using `epicsEvent`.

mark0n replied:
Seems like I forgot to commit my latest changes. Eliminating the first few `epicsThreadSleep`s was easy but I don't have any ideas how to get rid of the ones related to the cancel tests.

mark0n committed:
Ensure timers don't expire early

mark0n committed:
Eliminate a few epicsThreadSleep() calls

mark0n commented:
Superseded by #145.
